File: src/advent-of-code/2021/day-2/main.py
from pathlib import Path

file_path = "./input.txt"

def get_file_lines(file_path: str, encoding: str):
  file = Path(__file__).with_name(file_path)
  text = file.read_text(encoding)
  return text.splitlines()

def get_submarine_position(file_path: str):
  lines = get_file_lines(file_path, "utf-8")
  forward = 0
  depth = 0
  for line in lines:
    direction, value = parse_line(line)
    # i still use python 3.9.2 so no match statement here :[
    if direction == "forward":
      forward += value
    elif direction == "up":
      depth -= value
    elif direction == "down":
      depth += value
    else:
      raise ValueError(f"{direction} is an invalid directin")
  return forward * depth

def get_submarine_position_compound(file_path: str):
  aim = 0 
  depth = 0
  forward = 0
  lines = get_file_lines(file_path, "utf-8")
  for line in lines:
    direction, value = parse_line(line)
    if direction == "forward":
      forward += value
      depth += aim * value
    elif direction == "up":
      aim -= value
    elif direction == "down":
      aim += value
    else:
      raise ValueError(f"{direction} is an invalid directin")
  return forward * depth

def parse_line(line: str):
  parsed = line.split(" ")
  direction = parsed[0]
  value = int(parsed[1])
  return (direction, value)

if __name__ == "__main__":
  position = get_submarine_position(file_path)
  print(f"submarine position: {position}")
  position = get_submarine_position_compound(file_path)
  print(f"compound submarine position: {position}")

File: src/advent-of-code/2021/day-2/main.rb
def get_file_lines file_path
  IO.readlines(file_path, chomp: true)
end

def get_submarine_position lines
  depth = 0
  forward = 0
  for line in lines
    direction, value = parse_line line
    case direction
    when "forward"
     forward += value 
    when "up" 
      depth -= value
    when "down"
      depth += value
    else
      raise "#{direction} is an invalid direction" 
    end
  end
  depth * forward
end

def parse_line line
  direction, value = line.split(" ")
  return direction, value.to_i
end

INPUT_FILE_PATH = "./input.txt"

lines = get_file_lines INPUT_FILE_PATH
position = get_submarine_position lines
puts "Submarine Position: #{position}"
